Stage1 data cleaning

ZoeMZou opened the pull request:
### PR: Refinements for Script Structure & Data Processing
(Based on discussions on 28th January 2025)

**🔹 1. Structural Changes**
**📌 1.1 Naming Conventions**

- Removed `_stage1` from scripts to improve clarity and consistency.
- The old `stage1_data_cleaning.R` script was handling:
   1. Setting reference levels for categorical variables
   2. Applying quality assurance
   3. Applying inclusion/exclusion criteria

- The new structure separates these tasks into three dedicated functions:
   -   `fn-ref.R` → Sets reference levels
   -   `fn-qa.R` → Applies quality assurance
   -   `fn-inex.R`→ Applies inclusion/exclusion criteria

- The main script data_cleaning.R calls these functions, ensuring: Project-specific edits are made in function scripts and `data_cleaning.R` remains clear and workflow-focused.

**📌 1.2 Output Dataset Naming**

- The cleaned dataset output is now: input_{cohort}.rds

- The preprocessed dataset output is now:
input_{cohort}.rds → input_{cohort}_0.rds (to avoid YAML conflicts)

**🔹 2. `fn-ref.R` — Reference Level Settings**

**📌 2.1 Data Formatting Fixes**

- Some categorical variables were not factors, and some numeric variables were not numeric. I added code back to `preprocess_data.R` to ensure correct formatting:
[🔗 See fix in preprocess_data.R](https://github.com/opensafely/post-covid-respiratory/blob/cdeb9bc22aa023cecf816f0c4dc4f8ee4e20c8b8/analysis/preprocess/preprocess_data.R#L64-L72)

**📌 2.2 Code Deletions from Old Repo**
Index Date: Already correct—no correction needed.
[🔗 Old repo reference](https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/stage1_data_cleaning.R#L54-L58)
Deprivation (IMD): No need for re-categorization; new repo includes IMD (1–5) directly.
[🔗 Old repo reference](https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/stage1_data_cleaning.R#L97-L108)

**📌 2.3 Sex Category Adjustments**
Old repo: Only male (M) and female (F).
New repo: Four levels → female, male, intersex, unknown.
Fix:
Set non-male/non-female values to missing. Retain three levels: female, male, unknown.

**🔹 3. `fn-qa.R` — Quality Assurance Fixes**

**📌 3.1 Fixing "Date of Death" Message**
The original message incorrectly stated that NA values were being removed.
Fix: Message now correctly states:
"Quality assurance: Date of death is invalid (on or before 1/1/1900 or after current date)"

[🔗 Old repo reference](https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/stage1_data_cleaning.R#L179)

**📌 3.2 Fixing Missing Data Handling**
Issue: Patients with missing sex were  marked as missing for the entire record, which was unintended.
Fix: The new script preserves records and only marks sex as missing.
[🔗 Updated QA code](https://github.com/opensafely/post-covid-respiratory/blob/6adea6f962525a97912edef06ab0cbd8cbe046ca/analysis/data_cleaning/fn-qa.R#L35-L52)

**📌 3.3 Update to `qa_bin_prostate_cancer`**
Issue:
The current method introduced missing values due to logic operations with NA values. NA values were assigned to alive patients who had no recorded prostate cancer diagnosis but were affected by the logic.
The check for cause of death due to prostate cancer was unnecessary. Patients who died from prostate cancer before the index date are already excluded later in the study. Removing this check ensures all alive participants receive a valid classification (TRUE/FALSE) rather than NA.
[🔗 Updated method](https://github.com/opensafely/post-covid-respiratory/blob/6adea6f962525a97912edef06ab0cbd8cbe046ca/analysis/dataset_definition/variables_cohorts.py#L616-L623)

**🔹 4. `fn-inex.R` — Inclusion/Exclusion Criteria Fixes**

**📌 4.1 Code Corrections**
The new repo includes `inex_bin_alive`, so we can use it directly instead of recalculating.
[🔗 Old repo reference](https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/stage1_data_cleaning.R#L211-L216)
[🔗 New implementation](https://github.com/opensafely/post-covid-respiratory/blob/cdeb9bc22aa023cecf816f0c4dc4f8ee4e20c8b8/analysis/data_cleaning/stage_1_data_cleaning.R#L192-L197)

**📌 4.2 Variable Renaming**
`death_dat`e → `cens_date_death`
`has_follow_up_previous_6months` → `inex_bin_6m_reg`
`deregistration_date` → `cens_date_dereg`

**📌 4.3 Removal of Unnecessary Code**
Issue: The old repo had an unnecessary step for active registration at index.
Fix:
Removed redundant filtering for active registration, as inex_bin_6m_reg already ensures this. I rewrote the print message to reflect this.
[🔗 Old repo reference](https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/stage1_data_cleaning.R#L248-L253)
[🔗 Updated function](https://github.com/opensafely/post-covid-respiratory/blob/6adea6f962525a97912edef06ab0cbd8cbe046ca/analysis/data_cleaning/fn-inex.R#L38-L42)

**📌 4.4 Fix for Exclusion Criteria in the Vax Cohort**
Issue: Errors in the vax cohort exclusion criteria due to incorrect variable types.
Fix:
Redefined data types to ensure they are numeric before calculating vax_mixed.
[🔗 Updated fix](https://github.com/opensafely/post-covid-respiratory/blob/6adea6f962525a97912edef06ab0cbd8cbe046ca/analysis/data_cleaning/fn-inex.R#L85-L100)

**🔹 5. Other Updates**

**📌 5.1 Cause of Death Extraction Simplified**
A more efficient method for extracting cause of death has become available in OpenSAFELY.
Fix: Updated our code to align with the latest OpenSAFELY documentation.
[🔗 Updated function](https://github.com/opensafely/post-covid-respiratory/blob/6adea6f962525a97912edef06ab0cbd8cbe046ca/analysis/dataset_definition/variable_helper_functions.py#L153-L156)

ZoeMZou commented:
Hi @venexia,

This PR is now ready for your review. It runs successfully locally. Since the revisions in this script are quite detailed, I’ve listed the key changes at the beginning of the PR to make it easier to navigate and review.

Thank you very much.

Best wishes,
Zoe

venexia commented on analysis/dataset_definition/codelists.py:
```diff
@@ -423,25 +423,46 @@
 # Preexisting respiratory condition
 
-copd_ctv3 = codelist_from_csv(
+copd_ctv3_clinical = codelist_from_csv(
```
I thought we were removing the prefix '_clinical'?

ZoeMZou replied:
I thought we decided to resovle these naming issues in another PR? I could revise throughout in this PR if it is fine.

venexia commented on analysis/dataset_definition/codelists.py:
```diff
@@ -423,25 +423,46 @@
 # Respiratory outcome(s)
 
-pneumonia_snomed = codelist_from_csv(
+pneumonia_snomed_clinical = codelist_from_csv(
```
I thought we were removing the prefix '_clinical'?

venexia commented on analysis/dataset_definition/codelists.py:
```diff
@@ -423,25 +423,46 @@
+    column="ICD code"
+)
+
+pulmonary_fibrosis_snomed_clinical = codelist_from_csv(
```
I thought we were removing the prefix '_clinical'?

venexia reviewed (changes requested):
Hi @ZoeMZou. The PR looks good. Just a few things to address:
- Some of the QA criteria are out of date and do not match the current protocol
- The naming structure for the files through the pipeline is not very intuitive - we should agree an approach for this when we meet the post-covid-events team
- The currently generated dummy data does not have the variation to properly test this code - we should resolve this so we can double check we are removing the correct people (I did this for the first QA criterion but haven't gone through each criteria in this review)
- The formatting of the scripts is inconsistent and should be reviewed (we can automate  this using linting but I need to look up how to set it up)

ZoeMZou commented on analysis/data_cleaning/fn-qa.R:
```diff
@@ -0,0 +1,55 @@
+  consort[nrow(consort)+1,] <- c("Quality assurance: Date of death is invalid (on or before 1/1/1900 or after current date)",
+                                 nrow(input))
+  
+  print('Quality assurance: Pregnancy/birth codes for men')
```
Yes, individuals with missing sex should be included here. Only pregnancy/birth codes for men (HRT or COCP meds for men; prostate cancer codes for women) are excluded. We will handle the exclusion of individuals with missing sex in `fn-inex.R` to ensure accurate counts. Excluding them here would lead to inconsistencies in the reported number of missing sex cases.

ZoeMZou commented on analysis/data_cleaning/fn-qa.R:
```diff
@@ -0,0 +1,55 @@
+  consort[nrow(consort)+1,] <- c("Quality assurance: Year of birth is before 1793 or year of birth exceeds current date",
+                                 nrow(input))
+  
+  print('Quality assurance: Date of death is invalid (on or before 1/1/1900 or after current date)')
```
So are we revising protocol or remove the code here?

ZoeMZou commented on analysis/preprocess/preprocess_data.R:
```diff
@@ -112,7 +114,7 @@ df1[,colnames(df)[grepl("tmp_",colnames(df))]] <- NULL
 # Save input -------------------------------------------------------------------
 
-saveRDS(df1, file = paste0("output/input_",cohort_name,".rds"), compress = TRUE)
+saveRDS(df1, file = paste0("output/input_",cohort_name,"_0.rds"), compress = TRUE)
```
Decision from today's meeting (11th Feb 2025):

- We will not create an intermediate dataset from preprocess.

- After merging preprocess and data_cleaning, the only output will be `input_cohort_clean.rds`.

ZoeMZou commented:
Hi @venexia,
The PR runs okay locally and ready for your re-review. Thank you very much. Here is a summary of the changes:

**1. Overview of Changes**

Following in-depth discussions, the data cleaning process has been restructured under `analysis/dataset_clean/` to enhance clarity and efficiency:

- **Merged preprocess and stage_1_data_cleaning into a single script:**
   `dataset_clean.R` (now the main script handling both preprocessing and cleaning).

- **Introduced four functions for modular processing:**
   `fn-preproc.R` (data preprocessing, split into two functions vaccination overwriting does not work within function).
   `fn-ref.R` (reference level adjustments).
  `fn-inex.R` (inclusion/exclusion criteria).
   `fn-qa.R` (quality assurance checks).

**2. Key Revisions**

- **Eliminated intermediate dataset**
  Previously, preprocess generated an intermediate dataset, but this step is now removed.
  The only final dataset output will be:
📌 output/dataset_clean/input_{cohort}_clean.rds

-  **Order of Processing Adjusted**

   Inclusion criteria are now applied before quality assurance in aline with protocol.

**3. Additional Changes from Meeting (18/02/2024)**

- Added sub_cat_age to capture age categories.

- Integrated preprocessing updates from @harrytmason.

- Ensured outputs are written to the appropriate output/ subdirectories.

**4. File Naming Standardization**

Revised output filenames for better consistency and readability:

📂 Descriptive summaries
`output/describe/desc_preproc_{cohort}.txt`
`output/describe/desc_raw_{cohort}.txt`
`output/describe/desc_venn_{cohort}.txt`
📂 Datasets & consort
`output/dataset_clean/venn_{cohort}.rds`
`output/dataset_clean/consort_{cohort}.csv`
`output/dataset_clean/consort_{cohort}_midpoint6.csv`
**Final cleaned dataset:** `output/dataset_clean/input_{cohort}_clean.rds`

**5. Dummy Data Considerations**

Issue Identified:
The cleaned vax cohort contains zero records after applying the inclusion criterion:
"Inclusion criteria: Record of two vaccination doses prior to the study end date."
Reason:
Most patients in dummy data have missing vax_date_covid_1 or vax_date_covid_2.

venexia commented on analysis/create_project_actions.R:
```diff
@@ -14,6 +14,8 @@ defaults_list <- list(
 )
 cohorts <- c("prevax", "vax", "unvax")
+# describe_flag <- c("no_describe_print") #choose this to not print out describe*.txt files during dataset_clean.R
+describe_flag <- c("describe_print")   # choose this to print out describe*.txt files during dataset_clean.R
```
[Optional] Perhaps just use TRUE/FALSE?

venexia commented on analysis/dataset_clean/fn-qa.R:
```diff
@@ -0,0 +1,46 @@
+  consort[nrow(consort)+1,] <- c("Quality assurance: Year of birth exceeds current date",
+                                 nrow(input))
+  
+  print('Quality assurance: Date of death is invalid (on or before earliest expected date or after current date)')
```
I think the following is cleaner:
```
input <- input[(!is.na(input$cens_date_death) & 
(input$cens_date_death > as.Date(study_dates$earliest_expec)) & 
(input$cens_date_death <= format(Sys.Date(),"%Y-%m-%d"))) | 
is.na(input$cens_date_death),]
```

venexia replied:
Though the requirement about earliest expected date is not in the protocol?

ZoeMZou replied:
It was not in the protocol and I just found the `earliest_expec` date was not extracted from `study_dates.R` in `dataset_clen.R`. earliest_expec is set to 1900-01-01, which helps flag unrealistic death dates, however, since deaths before the index date are already excluded in inex. I feel we dont need it here.

venexia commented on analysis/dataset_clean/fn-qa.R:
```diff
@@ -0,0 +1,46 @@
+# Function to apply quality assurance
+qa <- function(input, consort, study_dates) {
+
+  print('Quality assurance: Year of birth is after year of death or patient only has year of death')
```
These are listed as two separate criteria in the protocol:
1. Remove individuals who are missing year of birth  
2. Remove individuals whose year of birth is after their year of death

ZoeMZou replied:
Oh, yes, it was in the old protocol but not in the ehr version. I’ve now updated the new one using track mode.

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,97 @@
+                              levels = c("Female", "Male", "Unknown"))
+  input$cov_cat_sex <- relevel(input$cov_cat_sex, ref = "Female")
+  
+  # Define cov_cat_age_group
```
As agreed at meeting 04/03/25, this will be removed from this script. (Apologies again for the back and forth on this decision!)

venexia reviewed (changes requested):
Hi @ZoeMZou. The code runs perfectly but some changes are required, as detailed in the code comments. They include:
- Making QA criteria match protocol
- Moving ref function to end of script
- Recoding conditions positively
- Improving dummy data so we can test all criteria

ZoeMZou commented on analysis/create_project_actions.R:
```diff
@@ -14,6 +14,8 @@ defaults_list <- list(
 )
 cohorts <- c("prevax", "vax", "unvax")
+# describe_flag <- c("no_describe_print") #choose this to not print out describe*.txt files during dataset_clean.R
+describe_flag <- c("describe_print")   # choose this to print out describe*.txt files during dataset_clean.R
```
I revised it to `describe_TRUE/FALSE`.

ZoeMZou commented on analysis/dataset_definition/variable_helper_functions.py:
You are right Venexia. This was used to create `cov_cat_age_group`. As we discussed on 4th March, we will remove this to table1.

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+inex <- function(input, cohort, vax_start_date, mixed_vax_threshold, start_date_delta) {
+
+  ## Specify consort table --------------------------------------------------------
+  print('Specify consort table')
```
I think cohortflow is a good name, another one I can think of is `**study_population_flow**`, but prefer `cohortflow` which is more concise.

venexia replied:
I think we could even just go for `flow`?

ZoeMZou replied:
Yes, agree, in the context of data cleaning, it defi refer to cohort flow.

ZoeMZou commented on analysis/dataset_clean/modify_dummy_data.R:
Sorry, Venxia, I’m a bit lost on this part. The current dummy data passed to this script actually includes people with the three conditions listed in the bullet points. Do you suggest to modify the dummy data to have more or less people with these condition, or would you prefer to discuss this further?

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+  
+  print('Inclusion criteria: Known sex at index')
+  
+  input <- input[!is.na(input$cov_cat_sex),] # removes NAs, if any
```
The positive codes were initially in fn-ref as the step before fn-inex, so any value other than male or female (e.g., intersex) was coded as NA in the fn-ref step.

Since we’re now moving fn-ref to the last step, as you suggested in a later comment, I’ll define the variables positively in this script (fn-inex) and adjust accordingly.

Just to double-check—does this look correct to you?

ZoeMZou commented on analysis/dataset_clean/fn-preproc.R:
As briefly mentioned in the description of this RP, the overwrite of dummy data codes couldn’t be incorporated into a function  (got error messages). To address this, the preprocess function was split into two at the point where we overwrite the dummy data.
https://github.com/opensafely/post-covid-respiratory/pull/80#:~:text=Introduced%20four%20functions%20for%20modular%20processing%3A%0Afn%2Dpreproc.R%20(data%20preprocessing%2C%20split%20into%20two%20functions%20vaccination%20overwriting%20does%20not%20work%20within%20function).

I can put them in two files with fn-preproc1 and fn-preproc2?

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+  ## Specify consort table --------------------------------------------------------
+  print('Specify consort table')
+
+  consort <- data.frame(Description = "Input", 
```
It has been removed to the begining of `dataset_clean.R`

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+  
+  print('Inclusion criteria: Known region at index')
+  
+  input <- input %>% mutate(strat_cat_region = as.character(strat_cat_region)) %>%
```
Yes, done.

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+  input <- input %>% mutate(strat_cat_region = as.character(strat_cat_region)) %>%
+    filter(strat_cat_region != "Missing")%>%
+    mutate(strat_cat_region = as.factor(strat_cat_region))
+  input$strat_cat_region <- relevel(input$strat_cat_region, ref = "East")
```
I have removed these to `fn-ref`.

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+  print('Inclusion criteria: Known IMD at index')
+  
+  input <- input[!is.na(input$cov_cat_imd),] # removes NAs, if any
+  input$cov_cat_imd <- ordered(input$cov_cat_imd, 
```
I have removed it to `fn-ref`.

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+
+    print('Inclusion criteria: Did not receive a vaccination prior to 08-12-2020 (i.e., the start of the vaccination program)')
+
+    input <- subset(input, input$vax_date_covid_1 >= vax_start_date & input$vax_date_covid_2 >= vax_start_date)
```
Yes. Done.

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+      rowwise() %>%
+      mutate(vax_mixed = sum(c_across(c(AZ_date, Moderna_date, Pfizer_date)), na.rm = TRUE)) %>%
+      ungroup() %>%
+      dplyr::filter(vax_mixed < 2)
```
Yes, that's right. I have removed ones we created temporarily 
```....
%>% select(-AZ_date, -Moderna_date, -Pfizer_date, -vax_mixed)

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+
+    input$prior_vax1 <- ifelse(input$vax_date_covid_1 <= input$index_date, 1,0)
+    input$prior_vax1[is.na(input$prior_vax1)] <- 0
+    input <- subset(input, input$prior_vax1 == 0) # Exclude people with prior vaccination
```
Yes, great, then we do not need `prior_vax1`

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+
+    print('Inclusion criteria: Index date is not before cohort end date - will remove anyone whose eligibility date + 84 days is after study end date (only those with unknown JCVI group)')
+
+    input <- input %>% filter (!is.na(index_date) & index_date <= end_date_exposure & index_date >= start_date_delta)
```
Thanks, Venexia! `subset()` seems more readable than `filter()` or `[]`, so I’ve updated the rest of the script to use `subset()` for consistency.

ZoeMZou commented on analysis/dataset_clean/dataset_clean.R:
```diff
@@ -0,0 +1,178 @@
+  ## Describe Venn diagram data
+input_venn <- preproc2(cohort, input_raw)$input_venn
+
+if (describe_flag == "describe_print") {
```
Yes, I have moved them into the function `preproc2`

ZoeMZou commented on analysis/dataset_clean/fn-inex.R:
```diff
@@ -0,0 +1,158 @@
+    print('Inclusion criteria: Record of two vaccination doses prior to the study end date')
+
+    input$vax_gap <- input$vax_date_covid_2 - input$vax_date_covid_1 #Determine the vaccination gap in days : gap is NA if any vaccine date is missing
+    input <- input[!is.na(input$vax_gap),] # Subset the fully vaccinated group
```
Thank you, Venexia. To ensure consistency and clarity, I have removed vax_gap throughout and explicitly coded the condition as follows:
https://github.com/opensafely/post-covid-respiratory/blob/7ddb21eba49a257dc28ecad3419bb020dc64b7a8/analysis/dataset_clean/fn-inex.R#L56-L80

ZoeMZou commented:
Hi @venexia,

Thanks so much for your comments—they’ve greatly improved the clarity and tidiness of the code, and I’ve learned a lot from them. I’ve implemented the suggested revisions, and the scripts now run successfully without errors. I’ve also provided detailed responses under each comment.

**A couple of key updates from my end:**

**1. Handling the `relevel()` error**
The current dummy data removes all patients with `East` after applying inclusion criteria. This initially caused an error:
```
Error in relevel.factor(input$strat_cat_region, ref = "East") :
  'ref' must be an existing level
```
I resolved this by explicitly setting factor levels regardless of whether "East" is present in the dataset. This prevents errors (Harry got last time, and I got it ealier) when releveling:https://github.com/opensafely/post-covid-respiratory/blob/25f388cc90ffebcd29b3c32e7fd87a3adb672469/analysis/dataset_clean/fn-ref.R#L46-L51

**2. `qa_num_birth_year` always set to 1975**
Harry and I identified the issue in fn-preproc1, which was causing `qa_num_birth_year` to always be 1975, that you proposed in earlier comments. We’ve temporarily commented out the problematic line, and the dates now look fine:
https://github.com/opensafely/post-covid-respiratory/blob/25f388cc90ffebcd29b3c32e7fd87a3adb672469/analysis/dataset_clean/fn-preproc1.R#L52

We should discuss whether this was intentional in the previous repository—if not, we can safely remove it?

**3. Next steps—Passing to @harrytmason for dummy data updates**
Harry and I had a meeting to investigate the step-by-step reduction in records. We identified the key points where records were removed and discussed adjusting the dummy data to maintain meaningful distributions. The branch is now with @harrytmason for further updates. Thanks, Harry!

harrytmason commented:
@venexia The code should now be ready for review

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,87 @@
+  # Handle missing values --------------------------------------------------------
+  print('Handle missing values')
+  
+  input$cov_cat_smoking <- replace(input$cov_cat_smoking, 
```
Code positively so replace everything with 'M' that is not a permitted value (E/N/S) for this variable.

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,87 @@
+  # Set reference level for variable: sub_cat_covidhospital -------------------
+  print('Set reference level for variable: sub_cat_covidhospital')
+  
+  input$sub_cat_covidhospital <- ordered(input$sub_cat_covidhospital, 
```
Why this order? Escalating severity would be "no_infection" > "non_hospitalised" > "hospitalised".

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,87 @@
+  
+  levels(input$cov_cat_sex) <- list("Female" = "female", "Male" = "male")
+  input$cov_cat_sex <- factor(input$cov_cat_sex, 
+                              levels = c("Female", "Male", "Unknown"))
```
Individuals with unknown sex should have been removed now that the ref function is at the end of the script.

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,87 @@
+  # Set reference level for variable: vax_cat_jcvi_group -------------------------
+  print('Set reference level for variable: vax_cat_jcvi_group')
+  
+  input$vax_cat_jcvi_group <- ordered(input$vax_cat_jcvi_group, 
```
Individuals with JCVI group 99 should have been removed now that the ref function is at the end of the script.

venexia commented on analysis/dataset_clean/fn-ref.R:
```diff
@@ -0,0 +1,87 @@
+                                          "Mixed" = "2", "South Asian" = "3", 
+                                          "Black" = "4", "Other" = "5")
+  
+  input$cov_cat_ethnicity <- ordered(input$cov_cat_ethnicity, 
```
This should use `relevel(ref = ...)` as ethnicity is not an ordinal variable.

venexia commented on analysis/dataset_clean/fn-modify_dummy.R:
```diff
@@ -0,0 +1,385 @@
+                  "03","02","01","99"), # 8.25 for each group, 1.6% for missing
+      size    = nrow(.),
+      replace = TRUE,
+      prob    = c(0.082,0.082,0.082,0.082,0.082,0.082,0.082,0.082,0.082,0.082,0.082,0.082, 0.016)))
```
```suggestion
      prob    = c(rep(0.082,12), 0.016)))
```

venexia commented on analysis/dataset_clean/fn-modify_dummy.R:
```diff
@@ -0,0 +1,385 @@
+                              ""), # 11% for each area, %1 Missing
+                  size    = nrow(.),
+                  replace = TRUE,
+                  prob    = c(0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.01))) %>%
```
Use `rep()` as demonstrated above for JCVI group.

venexia replied:
I think it is fine - it is not realistic, but variation is actually what is important here so that we expose the code to as many combinations of things as possible.

venexia commented on analysis/dataset_clean/fn-modify_dummy.R:
```diff
@@ -0,0 +1,385 @@
+                    as.Date("02-03-2023"), 
+                    .x))) %>%
+    
+    # mutate(modify_only_death = rbernoulli(nrow(.), p = 0.005)) %>%
```
Remove this code if not used - it is in the Git history if you need it.

harrytmason replied:
Just confirming to @ZoeMZou this is old code I don't need anymore and I'm good to remove it

venexia reviewed (changes requested):
Great work @ZoeMZou @harrytmason. A few more things to address before we merge to main. @ZoeMZou, are you happy to have a go at these?

ZoeMZou commented on analysis/dataset_clean/dataset_clean.R:
```diff
@@ -0,0 +1,158 @@
+# Save the dataset
+print('Saving dataset after Inclusion criteria')
+
+input <- input[, c(
```
Yes, I will keep the way consistent. Differences are to further restrict columns such as qa_ inex_

ZoeMZou commented:
Hi @venexia,
Thanks so much for your review! I’ve revised the code based on your comments and resolved the conversations accordingly. For those still open, I’ve added my thoughts and would appreciate your further feedback or confirmation.

As the code has been reformatted using the Air linting style, you may not be able to view specific commits easily via the “Files changed” tab. I’ve done my best to include permalinks in each response so you can directly see the changes.

**A couple of points that still need your input:**

1. In the vax-specific and unvax-specific inclusion criteria:

   Since vax_date_covid_1 is always before vax_date_covid_2 and vax_date_covid_3 by definition, do we prefer:

- a cleaner, shorter version that omits checks for vax_date_covid_2/3, or

- a more explicit version that includes them for clarity, even though the result will be the same?

2. Based on (1), I don’t think we need to artificially modify the dummy data to have vax_date_covid_2 earlier than vax_date_covid_1, since the flow diagram correctly shows 0 removed when the logic is valid. Could you confirm if you agree—and if so, let Harry know whether that dummy data tweak can be removed?

Thanks again!

venexia reviewed (commented):
Hi @ZoeMZou - really great work, this is looking really polished with the linting done! Three small things and then I think we are ready:
- [ ] Two of the QA criteria need to be updated to use last data collection date rather than `today`.
- [ ] In response to your query, I think we want the more explicit version that checks vax 2/3, which means the modified dummy data is fine as it is. I have made this decision because this is something that would need to be checked in other environments (it is fine in OS due to the way ehrQL extracts) so it makes this code more transportable for our team.
- [ ] There is a population size mismatch across files currently that should be resolved before we merge to main.

ZoeMZou commented:
Thanks very much @venexia. All done, EXCITING! 🎉

venexia reviewed (approved):
Great work! Let's get it running!
